sympy/matrices/common.py

```diff
@@ -2540,3 +2555,75 @@ def classof(A, B):
             return A.__class__
 
     raise TypeError("Incompatible classes %s, %s" % (A.__class__, B.__class__))
+
+
+def dotprodsimp(a, b, simplify=True):
+    """Sum-of-products with optional intermediate product simplification
+    targeted at the kind of blowup that occurs during summation of products.
+    Intended to reduce expression blowup during matrix multiplication or other
+    similar operations.
+
+    Parameters
+    ==========
+
+    a, b : iterable
+        These will be multiplied then summed together either normally or
+        using simplification on the intermediate products and cancelling at
+        the end according to the 'simplify' flag. The elements must already be
+        sympyfied and the sequences need not be of the same length, the shorter
+        will be used.
+
+    simplify : bool
+        When set intermediate and final simplification will be used, not set
+        will indicate a normal sum of products.
+    """
+
+    itra = iter(a)
+    itrb = iter(b)
+
+    try: # check for zero length
+        prod = next(itra)*next(itrb)
+    except StopIteration:
+        return S.Zero
```

Comment:
It isn't necessary to use iter, next or StopIteration here.

Reply:
This function can take an optional start parameter the same way that reduce does. In the matrices code it should just never be called with zero-length arguments since there is no need.